### PR TITLE
Add support for loading types out of RBIs and sig-annotated Ruby source files

### PR DESCRIPTION
**Pull request by AaronC81**
This is a huge change which ultimately adds `Parlour::TypeLoader`, a module which can be used to load type information back into Parlour's `RbiGenerator` objects from an RBI file or sig-annotated Ruby source code.

One can load individual files, or an entire Sorbet project by querying Sorbet's file table.

Currently supported:
  - Classes and modules, with modifiers (e.g. `abstract!`)
  - Methods
  - Attributes

There are some TODOs in this code for features which would be nice to have but aren't essential - I'll open issues for them once this is merged. I'll also update the README and add wiki entries once merged.

While this doesn't make any drastic changes to existing RBI generator functionality, I did have to make a few changes to be able to use the existing objects:
  - `RbiGenerator` is now entirely overriddable, which enabled the creation of the stub `DetachedRbiGenerator` used when loading.
  - The "parser" gem is now a dependency of Parlour.
  - `ClassNamespace` and `ModuleNamespace` instances can now be merged with `Namespace` instances.
  - A module and a namespace of the same name can now exist within the same namespace, and it will not be considered a conflict. (The Rainbow gem does this.)

### CODE REVIEW

File: lib/parlour/type_parser.rb [hunk 1]
```diff
@@ -0,0 +1,609 @@
...
+  # Parses Ruby source to find Sorbet type signatures.
+  class TypeParser
+    # Represents a path of indeces which can be traversed to reach a specific
+    # node in an AST.
+    class NodePath
+      extend T::Sig
+
+      sig { returns(T::Array[Integer]) }
+      # @return [Array<Integer>] The path of indeces.
+      attr_reader :indeces
```
Comment (connorshea) [inline lib/parlour/type_parser.rb]: I think this should be `indices`? As in the plural of index?

File: lib/parlour/type_parser.rb [hunk 2]
```diff
@@ -0,0 +1,609 @@
...
+    # any methods within.
+    #
+    # If the node directly represents several nodes, such as being a 
+    # (begin ...) node, they are all returned.
+    #
+    # @param [NodePath] path The path to the namespace definition. Do not pass
+    #   any of the other parameters to this method in an external call.
+    # @return [Array<RbiGenerator::RbiObject>] The objects the node at the path
+    #    represents, parsed into an RBI generator object.
+    sig { params(path: NodePath, is_within_eigen: T::Boolean).returns(T::Array[RbiGenerator::RbiObject]) }
```
Comment (connorshea) [inline lib/parlour/type_parser.rb]: What's eigen?

File: lib/parlour/type_parser.rb [hunk 3]
```diff
@@ -0,0 +1,609 @@
...
+    sig { params(node: T.nilable(Parser::AST::Node)).returns(T.nilable(String)) }
+    # Given an AST node, returns the source code from which it was constructed.
+    # If the given AST node is nil, this returns nil.
+    #
+    # @param [Parser::AST::Node, nil] node The AST node, or nil.
+    # @return [String] The source code string it represents, or nil.
+    def node_to_s(node)
+      return nil unless node
+
+      exp = T.unsafe(node).loc.expression
```
Comment (connorshea) [inline lib/parlour/type_parser.rb]: Does this still need to be unsafe since the RBI exists?

File: lib/parlour/type_parser.rb (hunk 2 above)
Comment (AaronC81) [reply lib/parlour/type_parser.rb]: I should probably rename this to `is_within_eigen_class` to be more clear.
The eigenclass is the Ruby object you're accessing when you do `class << self` from within a class.

File: sorbet/rbi/hidden-definitions/errors.txt [hunk 4]
```diff
@@ -37,7 +37,6 @@
 # wrong constant name <RESERVED_122>
 # wrong constant name <RESERVED_123>
 # wrong constant name <RESERVED_124>
-# wrong constant name <RESERVED_125>
```
[the remark this answers is not in the document]
Comment (AaronC81) [reply sorbet/rbi/hidden-definitions/errors.txt]: Good idea!

File: lib/parlour/type_parser.rb (hunk 1 above)
Comment (AaronC81) [reply lib/parlour/type_parser.rb]: Oops, you're right 🙃 Thanks!

File: lib/parlour/type_parser.rb (hunk 3 above)
Comment (AaronC81) [reply lib/parlour/type_parser.rb]: Ah, well spotted - Sorbet is still happy without the `T.unsafe`

**Pull request merged**